s2spy/time.py

```diff
@@ -304,3 +296,237 @@ def show(self) -> pd.DataFrame:
+                "Calerdar already exists and this overwrites the old calendar."
+            )
+        if period_block.target:
+            self._append_check(self._list_target, period_block)
+            self._list_target.append(period_block)
+            # count length
+            self._total_length_target += period_block.length + period_block.gap
+
+        else:
+            self._append_check(self._list_precursor, period_block)
+            self._list_precursor.append(period_block)
+            # count length
+            self._total_length_precursor += period_block.length + period_block.gap
+
+    def _append_check(self, list_period, period_block):
+        """Rules check for appending target precursor blocks."""
+        if not list_period:
+            if period_block.gap < 0:
+                raise ValueError(
+                    "First appended period block should not contain a negative gap."
+                )
+        else:
+            if (period_block.gap + list_period[-1].length) < 0:
+                raise ValueError(
+                    "Appended period block has negative gap larger than"
+                    + " the length of precedent period block."
+                )
+
+    def _map_year(self, year: int):
+        """Replace old map_year function"""
+        # flag the calendar as existed
+        self._end_indentifier = True
+
+        intervals_target = self._concatenate_periods(
+            year, self._list_target, self._total_length_target, True
+        )
+        intervals_precursor = self._concatenate_periods(
+            year, self._list_precursor, self._total_length_precursor, False
+        )
+
+        self.n_targets = len(intervals_target)
+        year_intervals = intervals_precursor[::-1] + intervals_target
+
+        # turn the list of intervals into pandas series
+        year_intervals = pd.Series(year_intervals[::-1], name=str(year))
+        year_intervals.index.name = "i_interval"
+        return year_intervals
+
+    def _concatenate_periods(self, year, list_periods, total_length, is_target):
+        # calculate start and end date
+        if is_target:
+            # anchor date is included
+            start_date = self._get_anchor(year)
+            end_date = start_date + pd.Timedelta(total_length, unit="D")
+            # generate date ranges
+            date_range = pd.date_range(start=start_date, end=end_date)
+        else:
+            # anchor date is excluded
+            start_date = self._get_anchor(year) - pd.Timedelta(
+                total_length + 1, unit="D"
+            )
+            end_date = self._get_anchor(year) - pd.Timedelta(1, unit="D")
+            # generate date ranges
+            date_range = pd.date_range(start=start_date, end=end_date)
+            date_range = date_range[::-1]
+        # generate intervals based on the building blocks
+        intervals = []
+        # pointer to the index of date for right boundary of interval
+        left_date_index = 0
```

Comment:
This pointer system storing the days from the anchor date as an `int` does not seem like the most efficient or robust way to do it. Why not base it completely off `pd.Timestamp` objects?

```py
intervals = []
date_pointer = anchor_timestamp # literal pd.timestamp, based on anchor date + year
for precursor in precursor_periods:
    left = date_pointer - pd.Timedelta(precursor.length + precursor.gap)
    intervals.append(pd.Interval(
        left=left,
        right=date_pointer,
        closed='left'))
    date_pointer = left

date_pointer = anchor_timestamp
for target in target_periods:
    # Similar but then build towards the right.
```

This would also make adding/subtracting those single days unnecessary.

Reply:
Yessss! Thanks @BSchilperoort ❤️! I should have thought about it! Brilliant idea! This really solves the problem for annoying index manipulation. Just implemented this. Now the concatenation looks very clean. 👍